# Bioinformatic training

<img align="left" width="300" src=./IM/LabLogo.png alt="Lab-Logo" />
Next-generation sequencing platforms have allowed researchers to generate whole-genome data from multiple individuals at an affordable cost while the optimization of extraction and library preparation protocols tailored for short and degraded DNA fragments have granted us the ability of retrieving genomic information from ancient remains. 

These technological advancements have revolutionized the fields of Evolutionary and Conservation genetics by enabling researchers to track genetic and demographic changes through time at an unprecedented level of detail. Ancient DNA (aDNA) research have not only dramatically changed our understanding of the speciation and domestication processes but also provided new insight on the complex evolutionary history of our species. 

Manipulating and analysing genomic data, however, requires a set of skills in bioinformatics that often prove to be a major barrier. Therefore we decided to create this course to teach our students and new lab members how to process and analyse next-generation sequencing data 
<br clear="left"/>

## Module 1 - Basic concepts of command line programing
In this module we will lay the fundations for the entire course. 

#### Session 1
- Accessing cloud server (ssh)
- Command Line Interface
- File System Structure
- Manipulating text files

#### Session 2
- Variables and arguments
- Shell Scripting
- Transferring files (sftp)
- Introduction to conda


## Module 2 - NGS data preparation for genomic analyses


#### Session 1
- Introduction to commonly used bioinformatic file formats
- Assessing the quality of sequencing results
- Aligning reads to reference genomes

#### Session 2
- Manipulating alignment files (bam)
- Estimating coverage, endogenous content, and percentage of duplicates

#### Session 3
- Parallelization strategies
- Variant calling

## Day 3 - Overview
Analysing genetic variation.

#### Session 1
- Understanding the .vcf file
- Calculating segregating sites, average number of pairwise differences, heterozygosity
- Hypothesis for the observed genetic diversity
- Discuss how to test these hypothesis

#### Session 2
- Generating phylogenetic trees
- Population structure with PCA
- Investigating population admixture

#### Session 3
- Visualising trees with iTOL
- Using R to visual PCA and ADMIXTURE results
- Discussion of Babirusa population structure

## Day 4 - Overview
Demographic analysis, inbreeding & load.

#### Session 1
- Prepping input files for PSMC
- Running and discussing PSMC

#### Session 2
- Running observational ROH detection method using PLINK
- Using R to visualise ROH observation-based and model-based results
- Discussion of different areas inbreeding extent

#### Session 3
- Running custom pipeline to calculate mutational load per individual
- Using R to plot mutational load per area
- Discussion of burden of mutation in different areas

## Day 5 - Overview
Advanced population genetics & Metabarcoding for biodiversity.

#### Session 1
- Introduction to f-statistics
- Detecting admixture
- Quatifying admixture proportions.

#### Session 2
- add some content here please
- two or three bullet point will suffice
- see my examples above
